src/server/notifications/WebSocketSubscription2021/WebSocket2021Emitter.ts

```diff
@@ -0,0 +1,90 @@
+import type { WebSocket } from 'ws';
+import type { SingleThreaded } from '../../../init/cluster/SingleThreaded';
+import { getLoggerFor } from '../../../logging/LogUtil';
+import { AsyncHandler } from '../../../util/handlers/AsyncHandler';
+import type { SetMultiMap } from '../../../util/map/SetMultiMap';
+import { WrappedSetMultiMap } from '../../../util/map/WrappedSetMultiMap';
+import { readableToString } from '../../../util/StreamUtil';
+import { setSafeInterval } from '../../../util/TimerUtil';
+import type { NotificationEmitter, NotificationEmitterInput } from '../NotificationEmitter';
+import type { SubscriptionStorage } from '../SubscriptionStorage';
+import type { WebSocket2021Handler, WebSocket2021HandlerInput } from './WebSocket2021Handler';
+
+/**
+ * Keeps track of the WebSockets that were opened for a WebSocketSubscription2021 subscription
+ * and uses them to emit events when necessary.
+ *
+ * SingleThreaded since the opened WebSocket might not be stored in the same thread
+ * as the one that emits a relevant event.
+ *
+ * `cleanupTimer` defines in minutes how often the stored WebSockets are closed
+ * if their corresponding subscription has expired.
+ * Defaults to 60 minutes.
+ * Although the WebSocket connection might still be open,
+ * if the subscription has expired no notifications will be sent.
+ */
+export class WebSocket2021Emitter extends AsyncHandler<NotificationEmitterInput | WebSocket2021HandlerInput>
+  implements NotificationEmitter, WebSocket2021Handler, SingleThreaded {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly storage: SubscriptionStorage;
+  private readonly webSockets: SetMultiMap<string, WebSocket>;
+
+  public constructor(storage: SubscriptionStorage, cleanupTimer = 60) {
+    super();
+
+    this.storage = storage;
+    this.webSockets = new WrappedSetMultiMap<string, WebSocket>();
+
+    const timer = setSafeInterval(this.logger,
+      'Failed to remove closed WebSockets',
+      this.closeExpiredSockets.bind(this),
+      cleanupTimer * 60 * 1000);
+    timer.unref();
+  }
+
+  public async handle(input: NotificationEmitterInput): Promise<void>;
+  public async handle(input: WebSocket2021HandlerInput): Promise<void>;
+  public async handle(input: NotificationEmitterInput | WebSocket2021HandlerInput): Promise<void> {
+    if (this.isWebSocket2021HandlerInput(input)) {
+      // Called as a WebSocket2021Handler: store the WebSocket
```

Comment:
Mmm this is a bit spooky. It's because we only have one `handle` method, right?
I think we should just break the `handler` pattern here then; and if we need it, have a small inner object for subscriptions?

Reply:
Darn I was hoping you were going to be "this is super cool!". Though about making a utility handler for this even 😉 .

But yes, it's because of the single `handle` call. I initially had a solution with 3 classes, one of them being the small inner object that just stored the websockets (which is the thing both these handlers need). But it felt silly to make the configuration more complicated for these tiny classes.

Reply:
I wonder if it's possible in TypeScript to have a generic class, where `handle` is part of the generic signature, such that the method names can be different 😅

But I really wouldn't overload `handle` like this. And if we really must (I don't think we do), we should dispatch it to 2 different internal methods then; not have the code inline.